### PR TITLE
added new Role entity, spring boot security basics, a few changesets …

### PR DESCRIPTION
**Pull request by annton-shakur**
added:
- SecurityConfigwith PasswordEncoder and SecurityFilterChain;
- Implementation of UserDetails interface in the User class;
- implementation for UserDetailsService;
- Role entity, RoleRepository;
- @PreAuthorize for all the BookController endpoints;
- Swagger for AuthController;

### CODE REVIEW

File: src/main/java/mate/academy/onlinebookstoreproject/controller/BookController.java [hunk 1]
```diff
@@ -27,31 +28,36 @@
...
     private final BookService bookService;
 
+    @PreAuthorize("hasRole('ROLE_USER')")
     @GetMapping
     @Operation(summary = "Get all books", description = "Get a pageable list of all the books")
     public List<BookDto> getAll(Pageable pageable) {
         return bookService.findAll(pageable);
     }
 
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Comment (vadymhrnk) [inline src/main/java/mate/academy/onlinebookstoreproject/controller/BookController.java]: It's better to use `hasAnyRole`.
If a user with the role `ADMIN` sends request on this and the previous endpoint, he'll get status code 403

File: src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+            tableName: users_roles
+            columns:
+              - column:
+                  name: role_id
+                  type: bigint
+                  constraints:
+                    nullable: false
+              - column:
+                  name: user_id
+                  type: bigint
```
Comment (vadymhrnk) [inline src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml]: I suggest to add this

                  constraints:
                    nullable: false
                    foreignKeyName: user_foreignKey
                    references: users(id)

to make both of this fields actual foreign keys

**Review by vadymhrnk — APPROVED** [review]
Good job! Left you 2 comments to consider. And a some of the `yaml` files missing endline

File: src/main/java/mate/academy/onlinebookstoreproject/config/SecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+                        auth -> auth
+                                .requestMatchers("/api/auth/**", "/error", "/swagger-ui/**")
+                                .permitAll()
+                                .anyRequest()
+                                .authenticated()
+                )
+                .httpBasic(Customizer.withDefaults())
+                .userDetailsService(userDetailsService)
+                .build();
+
```
Comment (bhdnchui) [inline src/main/java/mate/academy/onlinebookstoreproject/config/SecurityConfig.java]:
```suggestion
```

File: src/main/java/mate/academy/onlinebookstoreproject/controller/BookController.java (hunk 1 above)
Comment (bhdnchui) [reply src/main/java/mate/academy/onlinebookstoreproject/controller/BookController.java]: Great point!

File: src/main/java/mate/academy/onlinebookstoreproject/repository/UserRepository.java [hunk 4]
```diff
@@ -1,8 +1,9 @@
 package mate.academy.onlinebookstoreproject.repository;
 
+import java.util.Optional;
 import mate.academy.onlinebookstoreproject.model.User;
 import org.springframework.data.jpa.repository.JpaRepository;
 
 public interface UserRepository extends JpaRepository<User, Long> {
-    User findUserByEmail(String email);
+    Optional<User> findUserByEmail(String email);
```
Comment (bhdnchui) [inline src/main/java/mate/academy/onlinebookstoreproject/repository/UserRepository.java]: Rename to just `findByEmail`.

File: src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml (hunk 2 above)
Comment (bhdnchui) [reply src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml]: Great point!

File: src/main/resources/db/changelog/changes/05-create-roles-table.yaml [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: role
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
```
Comment (bhdnchui) [inline src/main/resources/db/changelog/changes/05-create-roles-table.yaml]: Please, add empty lines to your yaml files.
Make sure this red icon disappears
image

**Review by bhdnchui — CHANGES_REQUESTED** [review]
Good job!
Just a few points to consider.

**Review by bhdnchui — APPROVED** [review]
LGTM

**Pull request merged**